Add sq, cq, resources for queue and regex_dev, enqueue and dequeue functionalities and test

vuhuong opened the pull request:
This PR includes:
+ Add CQ create, destroy and arm helper APIs
+ Add SQ create and destroy helper APIs
+ Implement QP create callback
+ Add device and queue resources for enqueue dequeue
+ Add and enable enqueue and dequeue functions
+ Add test regex appliacation

chaudron committed:
net/i40e: re-program promiscuous mode on VF interface

During a kernel PF reset, this event is propagated to the VF.
The DPDK VF PMD will execute the reset task before the PF is done
with his. This results in the admin queue message not being responded
to leaving the port in "promiscuous" mode.

This patch makes sure the promiscuous mode is configured independently
of the current admin state.

Signed-off-by: Eelco Chaudron <echaudro@redhat.com>
Reviewed-by: Xiao Zhang <xiao.zhang@intel.com>

kaleshap86 committed:
app/testpmd: show MAC addresses added to a port

Patch adds a runtime function to display the unicast and
multicast MAC addresses added to a port.

Syntax:
	show port (port_id) macs|mcast_macs

Usage:
testpmd> show port 0 macs
Number of MAC address added: 1
  B0:26:28:7F:F5:C1
testpmd>
testpmd> show port 0 mcast_macs
Number of Multicast MAC address added: 0
testpmd>
testpmd> mac_addr add 0 B0:26:28:7F:22:33
testpmd> mac_addr add 0 B0:26:28:7F:22:34
testpmd> show port 0 macs
Number of MAC address added: 3
  B0:26:28:7F:F5:C1
  B0:26:28:7F:22:33
  B0:26:28:7F:22:34
testpmd>
testpmd> mac_addr remove 0 B0:26:28:7F:22:33
testpmd> show port 0 macs
Number of MAC address added: 2
  B0:26:28:7F:F5:C1
  B0:26:28:7F:22:34

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>
Reviewed-by: Ferruh Yigit <ferruh.yigit@intel.com>

Commit:
net/mlx5: separate DevX commands interface

The DevX commands interfaces is included in the mlx5.h file with a lot
of other PMD interfaces.

As an arrangement to make the DevX commands shared with different PMDs,
this patch moves the DevX interface to a new file called mlx5_devx_cmds.h.

Replace the DevX commands log mechanism from the mlx5 driver log
mechanism to the EAL log mechanism.

Signed-off-by: Matan Azrad <matan@mellanox.com>

hao022 committed:
net/mlx5: fix crash when setting hairpin queues

If configuring the number of tx/rx queue with rte_eth_dev_configure
to nr_queues + hairpin_nr_queues, and setting tx/rx queues to
nr_queues with rte_eth_tx/rx_queue_setup. But not configuring the
hairpin queues via rte_eth_tx/rx_hairpin_queue_setup.

When starting the netdev, there is a crash because of NULL accessing.

Fixes: cf5516696d77 ("ethdev: add hairpin queue")
Cc: stable@dpdk.org

Signed-off-by: Tonghao Zhang <xiangxia.m.yue@gmail.com>
Acked-by: Ori Kam <orika@mellanox.com>

hao022 committed:
net/mlx5: clean up redundant assignment

The type of txq has been assigned in mlx5_txq_hairpin_new.

Fixes: ae18a1ae9692 ("net/mlx5: support Tx hairpin queues")
Cc: stable@dpdk.org

Signed-off-by: Tonghao Zhang <xiangxia.m.yue@gmail.com>
Acked-by: Ori Kam <orika@mellanox.com>

Commit:
mlx5: prepare common library

A new Mellanox vdpa PMD will be added to support vdpa operations by
Mellanox adapters.

This vdpa PMD design includes mlx5_glue and mlx5_devx operations and big
part of them are shared with the net/mlx5 PMD.

Create a new common library in drivers/common for mlx5 PMDs.
Move mlx5_glue, mlx5_devx_cmds and their dependencies to the new mlx5
common library in drivers/common.

The files mlx5_devx_cmds.c, mlx5_devx_cmds.h, mlx5_glue.c,
mlx5_glue.h and mlx5_prm.h,  are moved as is from drivers/net/mlx5 to
drivers/common/mlx5.

Share the log mechanism macros.
Separate also the log mechanism to allow different log level control to
the common library.

Build files and version files are adjusted accordingly.
Includes lines are adjusted accordingly.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
mlx5: share the mlx5 glue reference

A new Mellanox vdpa PMD will be added to support vdpa operations by
Mellanox adapters.

Both, the mlx5 PMD and the vdpa mlx5 PMD should initialize the glue.

The glue initialization should be only one per process, so all the mlx5
PMDs using the glue should share the same glue object.

Move the glue initialization to be in common/mlx5 library to be
initialized by its constractor only once.

Signed-off-by: Matan Azrad <matan@mellanox.com>

kaleshap86 committed:
net/bnxt: refactor chip type selection

Use switch-case statement to replace if-else statement.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>

kaleshap86 committed:
net/bnxt: remove a log while deleting EM filters

Remove an unnecessary log while deleting EM filters.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>

kaleshap86 committed:
net/bnxt: use macro for getting default VNIC

BNXT_GET_DEFAULT_VNIC macro returns the default VNIC of the
function and fixed all occurrences of code to use the macro.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>

Commit:
mlx5: share mlx5 PCI device detection

Move PCI detection by IB device from mlx5 PMD to the common code.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
mlx5: share mlx5 devices information

Move the vendor information, vendor ID and device IDs from mlx5 PMD to
the common mlx5 file.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
drivers: introduce mlx5 vDPA driver

Add a new driver to support vDPA operations by Mellanox devices.

The first Mellanox devices which support vDPA operations are
ConnectX6DX and Bluefield1 HCA for their PF ports and VF ports.

This driver is depending on rdma-core like the mlx5 PMD, also it is
going to use mlx5 DevX to create HW objects directly by the FW.
Hence, the common/mlx5 library is linked to the mlx5_vdpa driver.

This driver will not be compiled by default due to the above
dependencies.

Register a new log type for this driver.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/ixgbe: enable jumbo frame for VF

Enable jumbo frame for VF by configuring DPDK PF.

Signed-off-by: Junyu Jiang <junyux.jiang@intel.com>
Acked-by: Wei Zhao <wei.zhao1@intel.com>

Commit:
net/iavf: add TSO offload use basic path

IAVF choices VECTOR TX path or Basic TX path by macro
IAVF_NO_VECTOR_FLAGS. TSO offload only processed by Basic TX path.

Fixes: 605c6f9ca52f ("net/avf: convert to new Rx and Tx offload API")
Cc: stable@dpdk.org

Signed-off-by: Tao Zhu <taox.zhu@intel.com>
Acked-by: Qi Zhang <qi.z.zhang@intel.com>

Commit:
net/ice: disable TSO offload in vector path

ICE choices VECTOR TX path or Basic TX path by macro
ICE_NO_VECTOR_FLAGS.
TSO offload only processed by Basic TX path.

Fixes: f88de4694d94 ("net/ice: support Tx SSE vector")
Cc: stable@dpdk.org

Signed-off-by: Tao Zhu <taox.zhu@intel.com>
Acked-by: Qi Zhang <qi.z.zhang@intel.com>

ciarapow committed:
app/testpmd: fix device mcast list error handling

The multicast set list function now has a return value, which is checked
by the calling functions. A rollback occurs on detection of failure, to
realign local config with the device config.

The error print statement in the function had included the port_id and
mc_addr_nb values in the wrong order, these are now swapped.

Fixes: 8fff667578a7 ("app/testpmd: new command to add/remove multicast MAC addresses")
Cc: stable@dpdk.org

Signed-off-by: Ciara Power <ciara.power@intel.com>
Reviewed-by: Ferruh Yigit <ferruh.yigit@intel.com>

luqiuwen committed:
net/fm10k: fix descriptor VLAN field filling in Tx

The VLAN field in the descriptor is not filled when ol_flags has
no VLAN flag. When the caller sends some packets with VLAN tags,
but other packets do not have VLAN tags, the behavior of the program
cannot be determined.

Fixes: 4b61d3bfa941 ("fm10k: add receive and tranmit")
Cc: stable@dpdk.org

Signed-off-by: Lu Qiuwen <luqiuwen@iie.ac.cn>
Acked-by: Xiao Wang <xiao.w.wang@intel.com>

shemminger committed:
net/netvsc: fix RSS offload flag

The change to add OFFLOAD_RSS_HASH broke use of netvsc PMD
with testpmd. The netvsc driver fails during configure step:

	Configuring Port 1 (socket 0)
	hn_dev_configure(): unsupported RX offload: 0x80000

Fixes: 5d308972954c ("ethdev: add mbuf RSS update as an offload")
Cc: stable@dpdk.org

Signed-off-by: Stephen Hemminger <stephen@networkplumber.org>

shemminger committed:
net/netvsc: disable before changing RSS parameters

In order to change RSS parameters, the host requires that RSS
is disable first.

Fixes: 92d23a57cafe ("net/netvsc: support configuring RSS parameters")
Cc: stable@dpdk.org

Signed-off-by: Donald Lee <Donald_Lee@trendmicro.com>
Signed-off-by: Stephen Hemminger <stephen@networkplumber.org>
Tested-by: Dexuan Cui <decui@microsoft.com>
Tested-by: Abhishek Marathe <abmarath@microsoft.com>

orikam committed:
regexdev: introduce regexdev subsystem

Even though there are some vendors which offer Regex HW offload, due to
lack of standard API, It is diffcult for DPDK consumer to use them
in a portable way.

This _RFC_ attempts to standardize the RegEx/DPI offload APIs for DPDK.

This RFC crafted based on SW Regex API frameworks such as libpcre and
hyperscan and a few of the RegEx HW IPs which I am aware of.

RegEx pattern matching applications:
* Next Generation Firewalls (NGFW)
* Deep Packet and Flow Inspection (DPI)
* Intrusion Prevention Systems (IPS)
* DDoS Mitigation
* Network Monitoring
* Data Loss Prevention (DLP)
* Smart NICs
* Grammar based content processing
* URL, spam and adware filtering
* Advanced auditing and policing of user/application security policies
* Financial data mining - parsing of streamed financial feeds
* Application recognition.
* Dmemory introspection.
* Natural Language Processing (NLP)
* Sentiment Analysis.
* Big data databse acceleration.
* Computational storage.

Request to review from HW and SW RegEx vendors and RegEx application
users to have portable DPDK API for RegEx.

The API schematics are based cryptodev, eventdev and ethdev existing
device API.

Signed-off-by: Jerin Jacob <jerinj@marvell.com>
Signed-off-by: Pavan Nikhilesh <pbhagavatula@marvell.com>
Signed-off-by: Ori Kam <orika@mellanox.com>

wangzk320 committed:
net/pcap: truncate packet if it is too large

Previously large packet would be dropped, instead now it is better to
keep it via truncating it.

Signed-off-by: Zhike Wang <wangzhike@jd.com>
Reviewed-by: Cian Ferriter <cian.ferriter@intel.com>
Acked-by: Ferruh Yigit <ferruh.yigit@intel.com>

Commit:
common/mlx5: expose vDPA DevX capabilities

Add the DevX capabilities for vDPA configuration and information of
Mellanox devices.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: support queues number operation

Support get_queue_num operation to get the maximum number of queues
supported by the device.

This number comes from the DevX capabilities.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: support features get operations

Add support for get_features and get_protocol_features operations.

Part of the features are reported by the DevX capabilities.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: glue null memory region allocation

Add support for rdma-core API to allocate NULL MR.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/ice: fix packet type table

According to comms package, index 94 should be MAC_IPV6_ICMP,
not MAC_IPV4_ICMP. This patch fixes the issue.

Fixes: 4cf29798b4c9 ("net/ice: fix packet type table")
Cc: stable@dpdk.org

Signed-off-by: Shougang Wang <shougangx.wang@intel.com>
Acked-by: Xiaolong Ye <xiaolong.ye@intel.com>

Commit:
common/mlx5: support DevX indirect mkey creation

Add option to create an indirect mkey by the current
mlx5_devx_cmd_mkey_create command.

Align the net/mlx5 driver usage in the above command.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: glue event queue query

The event queue is managed only by the kernel.

Add the rdma-core command in glue to query the kernel event queue
details.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: glue event interrupt commands

Add the next commands to glue in order to support interrupt event
channel operations associated to events in the EQ:
	devx_create_event_channel,
	devx_destroy_event_channel,
	devx_subscribe_devx_event,
	devx_subscribe_devx_event_fd,
	devx_get_event.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: glue UAR allocation

The isolated, protected and independent direct access to the HW by
multiple processes is implemented via User Access Region (UAR)
mechanism.
The UAR is part of PCI address space that is mapped for direct access to
the HW from the CPU.
UAR is comprised of multiple pages, each page containing registers that
control the HW operation.
UAR mechanism is used to post execution or control requests to the HW.
It is used by the HW to enforce protection and isolation between
different processes.

Add a glue command to allocate and free an UAR.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: add DevX command to create CQ

HW implements completion queues(CQ) used to post completion reports upon
completion of work request.

Used for Rx and Tx datapath.

Add DevX command to create a CQ.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: glue VAR allocation

Virtio access region(VAR) is the UAR that allocated for virtio emulation
access.

Add rdma-core operations to allocate and free VAR.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: add DevX virtio emulation commands

Virtio emulation offload allows SW to offload the I/O operations of a
virtio virtqueue, using the device, allowing an improved performance
for its users.
While supplying all the relevant Virtqueue information (type, size,
memory location, doorbell information, etc.). The device can then
offload the I/O operation of this queue, according to its device type
characteristics.

Some of the virtio features can be supported according to the device
capability, for example, TSO and checksum.

Add virtio queue create, modify and query DevX commands.

Signed-off-by: Matan Azrad <matan@mellanox.com>

kaleshap86 committed:
net/bnxt: restore MAC filters during reset recovery

Older Firmware could have state information such as
MAC Filters, VLAN settings etc configured by user.
But new Firmware is unaware of this state information
and as a result driver should restore these settings
during reset recovery.

This patch restores the user configured mac addresses
prior to hot FW upgrade or FW error.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

kaleshap86 committed:
net/bnxt: restore VLAN filters during reset recovery

Older Firmware could have state information such as
MAC Filters, VLAN settings etc configured by user.
But new Firmware is unaware of this state information
and as a result driver should restore these settings
during reset recovery.

This patch restores the user configured vlan settings
prior to hot FW upgrade or FW error.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

kaleshap86 committed:
net/bnxt: fix request for hot reset support

During the registration process, the drivers must publish
"HOT_RESET_SUPPORTED" capability to the core firmware only
if core firmware is capable of Hot Reset.

Fixes: df6cd7c1f73a ("net/bnxt: handle reset notify async event from FW")
Cc: stable@dpdk.org

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>

kaleshap86 committed:
net/bnxt: remove unnecessary macro for unused variables

Remove "__rte_unused" instances that are not required.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Rahul Gupta <rahul.gupta@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>

kaleshap86 committed:
net/bnxt: reset filter indices on free

Fix bnxt_free_filter() to ensure that the filter indices
are reset correctly when a filter is freed.

This change helps in avoiding few duplicate code as well.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Somnath Kotur <somnath.kotur@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

kaleshap86 committed:
net/bnxt: add a field for FW capabilities

The purpose of this patch is used to add a new field "fw_cap"
in bnxt structure. The new "fw_cap" field store all the capabilities
reported by the FW. This information will help us to determine
the device capablities when registering feature support with FW.

Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

Commit:
net/i40e: fix port close in FreeBSD

FreeBSD OS doesn't support igb_uio interrupt, so it will fail when
unregister the interrupt callback in port close. We can fix the issue by
passing -ENOENT when check the return value of unregister interrupt
callback function.

Fixes: ac89d46096d5 ("net/i40e: release port upon close")
Cc: stable@dpdk.org

Signed-off-by: Chenxu Di <chenxux.di@intel.com>
Acked-by: Xiaolong Ye <xiaolong.ye@intel.com>

0day-trouble-maker committed:
maintainers: adjust maintainership of intel tree

Currently I am the main committer of dpdk-next-net-intel tree.

Signed-off-by: Xiaolong Ye <xiaolong.ye@intel.com>
Acked-by: Qi Zhang <qi.z.zhang@intel.com>
Acked-by: Ferruh Yigit <ferruh.yigit@intel.com>

Commit:
vdpa/mlx5: prepare memory regions

In order to map the guest physical addresses used by the virtio device
guest side to the host physical addresses used by the HW as the host
side, memory regions are created.

By this way, for example, the HW can translate the addresses of the
packets posted by the guest and to take the packets from the correct
place.

The design is to work with single MR which will be configured to the
virtio queues in the HW, hence a lot of direct MRs are grouped to single
indirect MR.

Create functions to prepare and release MRs with all the related
resources that are required for it.

Create a new file mlx5_vdpa_mem.c to manage all the MR related code
in the driver.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
maintainers: update for OCTEONTX

Harman thankfully accepted to replace myself as maintainer for OCTEONTX
ethdev PMD.

Signed-off-by: Jerin Jacob <jerinj@marvell.com>
Acked-by: Harman Kalra <hkalra@marvell.com>
Acked-by: Ferruh Yigit <ferruh.yigit@intel.com>

Commit:
mlx5: share CQ entry check

The CQE has owner bit to indicate if it is in SW control or HW.

Share a CQE check for all the mlx5 drivers.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: prepare completion queues

As an arrangement to the vitrio queues creation, a CQ should be created
for the virtio queue.

The design is to trigger an event for the guest and for the vdpa driver
when a new CQE is posted by the HW after the packet transition.

This patch add the basic operations to create and destroy CQs and to
trigger the CQE events when a new CQE is posted.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: handle completions

In order to free resources in the CQ and to allow to the HW to send
packets from the guest, the CQ should be polled.

In order to poll the CQ we need to trigger an interrupt for each new
CQE posted by the HW.

Register interrupt handler to poll and arm a CQ when completion event
was raised by the HW.

Signed-off-by: Matan Azrad <matan@mellanox.com>

bruce-richardson committed:
net/ipn3ke: fix meson build

The ipn3ke driver and the ifpga driver had circular dependencies on each
other, meaning that neither could be built with meson. Fix this by:

* setting the build order of net and raw to match that in makefiles
* removing the dependency from the net driver on the rawdev one
* fixing the dependency names in the rawdev driver meson.build file

Fixes: 8418c92811b4 ("net/ipn3ke: remove configuration for i40e port bonding")
Fixes: e1defba4cf66 ("raw/ifpga/base: support device tree")
Cc: stable@dpdk.org

Signed-off-by: Bruce Richardson <bruce.richardson@intel.com>
Acked-by: Rosen Xu <rosen.xu@intel.com>

jiaqimix committed:
net/iavf: fix Rx total stats

Rx total stats is the total number of successfully received packets,
so exclude the number of rx_discards for Rx total stats.

Fixes: f4a41a6953af ("net/avf: support stats")
Cc: stable@dpdk.org

Signed-off-by: Jiaqi Min <jiaqix.min@intel.com>
Acked-by: Jingjing Wu <jingjing.wu@intel.com>

Commit:
vdpa/mlx5: prepare virtio queues

The HW virtq object represents an emulated context for a VIRTIO_NET
virtqueue which was created and managed by a VIRTIO_NET driver as
defined in VIRTIO Specification.

Add support to prepare and release all the basic HW resources needed
the user virtqs emulation according to the rte_vhost configurations.

This patch prepares the basic configurations needed by DevX commands to
create a virtq.

Add new file mlx5_vdpa_virtq.c to manage virtq operations.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: support stateless offloads

Add support for the next features in virtq configuration:
	VIRTIO_F_RING_PACKED,
	VIRTIO_NET_F_HOST_TSO4,
	VIRTIO_NET_F_HOST_TSO6,
	VIRTIO_NET_F_CSUM,
	VIRTIO_NET_F_GUEST_CSUM,
	VIRTIO_F_VERSION_1,

These features support depends in the DevX capabilities reported by the
device.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/ixgbe: fix port can not link up in FreeBSD

In FreeBSD environment, nic_uio drivers do not support interrupts,
rte_intr_callback_register() will fail to register interrupts.
We can not make link status to change from down to up by interrupt
callback. So we need to wait for the controller to acquire link
when ports start. Through multiple tests, 5s should be enough.

Fixes: b9bd0f09fa15 ("ethdev: fix link status query")
Cc: stable@dpdk.org

Signed-off-by: Lunyuan Cui <lunyuanx.cui@intel.com>
Acked-by: Xiaolong Ye <xiaolong.ye@intel.com>
Acked-by: Konstantin Ananyev <konstantin.ananyev@intel.com>

Commit:
common/mlx5: allow type configuration for DevX RQT

Allow virtio queue type configuration in the RQ table.
The needed fields and configuration was added.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: add TIR fields constants

The DevX TIR object configuration should get L3 and L4 protocols
expected to be forwarded by the TIR.

Add the PRM constant values needed to configure the L3 and L4 protocols.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: add basic steering configurations

Add a steering object to be managed by a new file mlx5_vdpa_steer.c.

Allow promiscuous flow to scatter the device Rx packets to the virtio
queues using RSS action.

In order to allow correct RSS in L3 and L4, split the flow to 7 flows
as required by the device.

Signed-off-by: Matan Azrad <matan@mellanox.com>

jackmin committed:
net/mlx5: fix multiple flow table hash list

The eth devices which share one ibv device only need one hash list of
flow table.

Currently, flow table hash list is created per each eth device
whatever whether they share one ibv device or not.

If the devices share one ibv device, the previously created hash list
will become dangle because the pointer point to (sh->flow_tbls) is
overwritten by the later created hast list.

To fix this, just don't create hash list if it is already created.

Fixes: 54534725d2f3 ("net/mlx5: fix flow table hash list conversion")
Cc: stable@dpdk.org

Reported-by: Zhike Wang <wangzhike@jd.com>
Signed-off-by: Xiaoyu Min <jackmin@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: support queue state operation

Add support for set_vring_state operation.

Using DevX API the virtq state can be changed as described in PRM:
	enable - move to ready state.
	disable - move to suspend state.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: map doorbell

The HW support only 4 bytes doorbell writing detection.
The virtio device set only 2 bytes when it rings the doorbell.

Map the virtio doorbell detected by the virtio queue kickfd to the HW
VAR space when it expects to get the virtio emulation doorbell.

Use the EAL interrupt mechanism to get notification when a new event
appears in kickfd by the guest and write 4 bytes to the HW doorbell space
in the notification callback.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/ixgbe: fix link status

The link status for ixgbevf is not correct when PF link up.
IXGBE_ESDP register is only used when media type is fiber.

Fixes: 1ca05831b9be ("net/ixgbe: fix link status")
Cc: stable@dpdk.org

Signed-off-by: Lunyuan Cui <lunyuanx.cui@intel.com>
Acked-by: Xiaolong Ye <xiaolong.ye@intel.com>

Commit:
vdpa/mlx5: support live migration

Add support for live migration feature by the HW:
	Create a single Mkey that maps the memory address space of the
		VHOST live migration log file.
	Modify VIRTIO_NET_Q object and provide vhost_log_page,
		dirty_bitmap_mkey, dirty_bitmap_size, dirty_bitmap_addr
		and dirty_bitmap_dump_enable.
	Modify VIRTIO_NET_Q object and move state to SUSPEND.
	Query VIRTIO_NET_Q and get hw_available_idx and hw_used_idx.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: support close and config operations

Support dev_conf and dev_conf operations.
These operations allow vdpa traffic.

Signed-off-by: Matan Azrad <matan@mellanox.com>

shemminger committed:
ethdev: fix flow API doxygen comment

Missing asterisk so that comment is not seen by doxygen.

Fixes: 9a2f44c76207 ("ethdev: add flow tag")
Cc: stable@dpdk.org

Signed-off-by: Stephen Hemminger <stephen@networkplumber.org>
Acked-by: Viacheslav Ovsiienko <viacheslavo@mellanox.com>

dekelp committed:
net/mlx5: remove redundant define of LRO masks

Bit-masks MLX5_FLOW_LAYER_IPV4_LRO and MLX5_FLOW_LAYER_IPV6_LRO were
added during work on LRO.
They are not used anywhere in current code.

This patch removes these redundant definitions.

Signed-off-by: Dekel Peled <dekelp@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

dekelp committed:
net/mlx5: add define of LRO segment chunk size

Maximal size of coalesced LRO segment is set in TIR attributes as
number of chunks of size 256 bytes each.
Current implementation uses the hardcoded value 256 in several places.

This patch adds a definition for this value, and uses this definition
in all relevant places.
A debug message is added to clearly notify the actual configured size.

Signed-off-by: Dekel Peled <dekelp@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

Commit:
drivers: introduce vDPA class

The vDPA (vhost data path acceleration) drivers provide support for
the vDPA operations introduced by the rte_vhost library.

Any driver which provides the vDPA operations should be moved\added to
the vdpa class under drivers/vdpa/.

Create the general files for vDPA class in drivers and in documentation.

The management tree for vDPA drivers is
git://dpdk.org/next/dpdk-next-virtio.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
drivers: move ifc driver to the vDPA class

A new vDPA class was recently introduced.

IFC driver implements the vDPA operations, hence it should be moved to
the vDPA class.

Move it.

Signed-off-by: Matan Azrad <matan@mellanox.com>

dekelp committed:
net/mlx5: fix setting of Rx hash fields

Rx hash fields were copied from input parameter into TIR attributes
directly, with no translation. As result the copied value was wrong.

This patch adds translation of value from input bitmap to the
appropriate format.

Fixes: dc9ceff73c99 ("net/mlx5: create advanced RxQ via DevX")
Cc: stable@dpdk.org

Signed-off-by: Dekel Peled <dekelp@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

dekelp committed:
net/mlx5: support RSS on SRC or DST fields only

This patch implements the feature described in RFC [1], adding
support of RSS action on L3 and/or L4 source or destination only.

[1] http://mails.dpdk.org/archives/dev/2019-December/152796.html

Signed-off-by: Dekel Peled <dekelp@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

tingxu1 committed:
net/ice: add device IDs

This patch added new device IDs for C822N.

Signed-off-by: Ting Xu <ting.xu@intel.com>
Acked-by: Xiaolong Ye <xiaolong.ye@intel.com>

Commit:
net/mlx5: skip probing in vDPA mode

Both, net/mlx5 PMD and vdpa/mlx5 drivers can probe the same Mellanox
devices.

Skip net/mlx5 PMD probing while the device is in vDPA mode selected by
the device devargs provided by the user: vdpa=1.

Signed-off-by: Matan Azrad <matan@mellanox.com>

dekelp committed:
net/mlx5: fix item flag on GENEVE item validation

On validation of GENEVE item, the wrong flag is used.
This patch sets the correct bit in item_flags bitmap.

Fixes: e59a5dbcfd07 ("net/mlx5: add flow match on GENEVE item")
Cc: stable@dpdk.org

Signed-off-by: Dekel Peled <dekelp@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

viacheslavo committed:
net/mlx5: fix matcher metadata register c0 field setup

The metadata register c0 field in the matcher might be split
into two independent fields - the source vport index and META
item value. These fields have no permanent assigned bits, the
configuration is queried from the kernel drivers.

MLX5_SET configures the specified 32-bit field as whole entity.
For metadata register c0 we should take into account the provided
mask in order to configure the specified subfield bits only.

Fixes: acfcd5c52f94 ("net/mlx5: update meta register matcher set")
Cc: stable@dpdk.org

Signed-off-by: Viacheslav Ovsiienko <viacheslavo@mellanox.com>
Acked-by: Matan Azrad <matan@mellanox.com>

Commit:
doc: add vDPA feature table

Add vDPA devices features table and explanation.

Any vDPA driver can add its own supported features by ading a new ini
file to the features directory in doc/guides/vdpadevs/features.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/mlx5: separate Netlink commands interface

The Netlink commands interfaces is included in the mlx5.h file with a
lot of other PMD interfaces.

As an arrangement to make the Netlink commands shared with different
PMDs, this patch moves the Netlink interface to a new file called
mlx5_nl.h.

Move non Netlink pure vlan commands from mlx5_nl.c to the
mlx5_vlan.c.

Rename all the Netlink commands and structure to use prefix mlx5_nl.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
net/mlx5: reduce Netlink commands dependencies

As an arrangment for Netlink command moving to the common library,
reduce the net/mlx5 dependecies.

Replace ethdev class command parameters.

Improve Netlink sequence number mechanism to be controlled by the
mlx5 Netlink mechanism.

Move mlx5_nl_check_switch_info to mlx5_nl.c since it is the only one
which uses it.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
mlx5: share Netlink commands

Move Netlink mechanism and its dependecies from net/mlx5 to common/mlx5
in order to be ready to use by other mlx5 drivers.

The dependencies are BITFIELD defines, the ppc64 compilation workaround
for bool type and the function mlx5_translate_port_name.

Update build mechanism accordingly.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: support ROCE disable through Netlink

Add new 4 Netlink commands to support enable/disable ROCE:
        1. mlx5_nl_devlink_family_id_get to get the Devlink family ID of
           Netlink general command.
        2. mlx5_nl_enable_roce_get to get the ROCE current status.
        3. mlx5_nl_driver_reload - to reload the device kernel driver.
        4. mlx5_nl_enable_roce_set - to set the ROCE status.

When the user changes the ROCE status, the IB device may disappear and
appear again, so DPDK driver should wait for it and to restart itself.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
vdpa/mlx5: disable ROCE

In order to support virtio queue creation by the FW, ROCE mode
should be disabled in the device.

Do it by netlink which is like the devlink tool commands:
	1. devlink dev param set pci/[pci] name enable_roce value false
	   cmode driverinit
    	2. devlink dev reload pci/[pci]
Or by sysfs which is like:
	echo 0 >  /sys/bus/pci/devices/[pci]/roce_enable

The IB device is matched again after ROCE disabling.

Signed-off-by: Matan Azrad <matan@mellanox.com>

Commit:
common/mlx5: Add MMO and regexp structs/opcodes

Added General purpose PRM MMO structs, and regex specifix structs.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

Commit:
common/mlx5: add mlx5 regex command structs

Add regex commands structs to support regex.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

Commit:
regex/mlx5: add Makefiles

Add Makefiles to the regex tree.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

Commit:
regex/mlx5: add mlx5 regex src file

mlx5_regex.c will abstract API to FW/HW regex operations.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

Commit:
common/mlx5: add support for regex capability query

Update hca cap struct and common query hca cap function.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

Commit:
regex/mlx5: add database commands support

Add regex database commands:
 - Stop engine
 - Set DB umem
 - Resume

Database init flow per engine:
1. Set database umem.
2. Resume.

Database update flow:
1. Stop engine.
1. Set database.
3. Resume.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

orikam committed:
regexdev: add regex core h file

This commit introduce the rte_regexdev_core.h file.
This file holds internal structures and API that are used by
the regexdev.

Signed-off-by: Ori Kam <orika@mellanox.com>

Commit:
regex/mlx5: Add RXP register read/write commands

RXP is configured through a series or registers write/read
operations.
These registers are also used for statistics.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

orikam committed:
regexdev: add regexdev core functions

Signed-off-by: Ori Kam <orika@mellanox.com>
Signed-off-by: Parav Pandit <parav@mellanox.com>

Commit:
regex/mlx5: Add log utils

Add the DRV_LOG macro which should be used for error prints.

Signed-off-by: Yuval Avnery <yuvalav@mellanox.com>

orikam committed:
config/common: add define for mlx5 regex pmd

Signed-off-by: Ori Kam <orika@mellanox.com>

paravmellanox committed:
regex/mlx5: Register unregister mlx5 regex device

When a PCI device is capable of regex feature, register a regex device
with regex subsystem.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
regex/mlx5: Add RXP register definitions

Signed-off-by: Parav Pandit <parav@mellanox.com>

orikam committed:
regexdev: implement regex rte level functions

This commit implements all the RegEx public API.

Signed-off-by: Ori Kam <orika@mellanox.com>

Commit:
regex/mlx5: Enable regex mlx5 pmd build and link

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

paravmellanox committed:
regex/mlx5: Setup device PD and EQN

Setup devie PD and EQN to operate the device.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
app/regex: Add test regex appliacation

Add test application to test RTE regex APIs.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
regex/mlx5: Extend probe to initialize database memory

Extended probe routine to register database memory with the
device.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
regex/mlx5: Implement devinfo get callback

Implement devinfo get callback routine.

Signed-off-by: Parav Pandit <parav@mellanox.com>

Commit:
regex/mlx5: Add Memory Region and its management data structures

Add Memory Region(MR), its cache and control management data
structures.

Move mlx5_regex_priv to header file and add extra field for
MRs cache for each regex device.

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

Commit:
regex/mlx5: Implement MR registration and management

Implement MR APIs to:
 . Search for a LKey of a given virtual address in per-queue local
   cache and global cache shared by multiple queue on the same
   rte_regex_dev.
 . Create a new MR for a missing virtual address, register entire
   virtually contiguous memory chunk around the address.
 . Release all MRs and resources in the shared device context.
 . Initialize <lkey,addr> cache B-tree table, and insert/lookup
   operations.

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

Commit:
regex/mlx5: Add and enable enqueue and dequeue functions

Implement and enable REGEX device's enqueue and dequeue APIs.

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

paravmellanox committed:
regex/mlx5: Implement QP create callback

Support QP creation callback and necessary device configure callback.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
common/mlx5: Add CQ create, destroy and arm helper APIs

Add APIs to create, destroy and arm a CQ.

Signed-off-by: Parav Pandit <parav@mellanox.com>

paravmellanox committed:
common/mlx5: Add SQ create and destroy helper APIs

Signed-off-by: Parav Pandit <parav@mellanox.com>

Commit:
regex/mlx5: Add device and queue resources for enqueue dequeue

Initalize btree cache per device.
For each queue:
. Initalize btree cache per queue.
. Allocate job data structure with metadata and response buffers
  to handle nb_desc.

Signed-of-by: Vu Pham <vuhuong@mellanox.com>

orikam commented on drivers/regex/mlx5/mlx5_regex_queue.c:
```diff
@@ -0,0 +1,261 @@
+			uint32_t mmo_control_31_0, uint32_t lkey,
+			uintptr_t address)
+{
+	seg->mmo_control_31_0 = htobe32(mmo_control_31_0);
```
should be changed to rte_cpu_to_be_32
same for all 3 lines

orikam commented on drivers/regex/mlx5/mlx5_regex_queue.c:
```diff
@@ -0,0 +1,261 @@
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					job->m_iov.buf_iova);
+	mlx5dv_set_data_seg(&tmp_dseg, job->m_iov.buf_size,
+			    lkey, job->m_iov.buf_iova);
```
we should create those function in DPDK.
buf_iova should be changed to buf addr

orikam reviewed (changes requested):
I will review more later

Commit:
common/mlx5: Move memory management codes to common driver

Move memory management to common driver.
Changes net pmd and regex drivers to use MR in common driver.

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

Commit:
common/mlx5: Add common primary/secondary processes handling codes

Move common mp handling codes from net pmd to common drive.

Signed-off-by: Vu Pham <vuhuong@mellanox.com>

yuvalavn commented on drivers/regex/mlx5/mlx5_regex_queue.c:
```diff
@@ -0,0 +1,263 @@
+			    lkey, (uintptr_t)BUF_ADDR(&job->resp_iov));
+	/* Ring doorbell */
+	rte_io_wmb();
+	q->sq.db_rec[MLX5_SND_DBR] = rte_cpu_to_be_32(q->sq.pi & MLX5_QP_CI_MASK);
```
Hi vu
this a bug
please see PRM
need to be the next empty wqbb and not wqe index

yuvalavn commented on drivers/regex/mlx5/mlx5_regex_queue.c:
```diff
@@ -0,0 +1,263 @@
+#ifdef RTE_ARCH_64
+	*(uint64_t *)q->sq.db_addr = *(uint64_t *)cseg;
+#else
+	*(uint32_t *)q->sq.db_addr = *(uint64_t *)cseg;
```
Every write need to be alternated between odd and even on the uar
please see PARM

yuvalavn replied:
this comment relates to blueflame

yuvalavn commented on drivers/regex/mlx5/mlx5_regex_queue.c:
```diff
@@ -0,0 +1,263 @@
+	q->sq.db_rec[MLX5_SND_DBR] = rte_cpu_to_be_32(q->sq.pi & MLX5_QP_CI_MASK);
+	rte_wmb();
+#ifdef RTE_ARCH_64
+	*(uint64_t *)q->sq.db_addr = *(uint64_t *)cseg;
```
We should use blueflame for performance
It means writing the whole wqe to the BAR not just the ctrl segment

yuvalavn replied:
Should be
*(volatile uint64_t *)cseg;

Commit:
regex/mlx5: switch to mbuf

Signed-off-by: Vu Pham <vuhuong@mellanox.com>
